File: python/332_ReconstructItinerary.py
from unittest import TestCase
# https://leetcode.com/problems/reconstruct-itinerary/
import collections


class ReconstructItinerary(TestCase):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        tos = collections.defaultdict(list)
        for fr, to in sorted(tickets)[::-1]:
            tos[fr].append(to)
        ret = []

        def dfs(fr):
            while tos[fr]:
                dfs(tos[fr].pop())
            ret.append(fr)
        dfs("JFK")
        return ret[::-1]

    def test1(self):
        self.assertEqual(["JFK", "MUC", "LHR", "SFO", "SJC"], self.findItinerary(
            [["MUC", "LHR"], ["JFK", "MUC"], ["SFO", "SJC"], ["LHR", "SFO"]]
        ))

    def test2(self):
        self.assertEqual(["JFK","ATL","JFK","SFO","ATL","SFO"], self.findItinerary(
            [["JFK", "SFO"], ["JFK", "ATL"], ["SFO", "ATL"], ["ATL", "JFK"], ["ATL", "SFO"]]
        ))

    def test3(self):
        self.assertEqual(["JFK","NRT","JFK","KUL"], self.findItinerary(
            [["JFK", "KUL"], ["JFK", "NRT"], ["NRT", "JFK"]]
        ))

File: kotlin/src/main/kotlin/332.ReconstructItinerary.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/reconstruct-itinerary/
import java.util.*


class ReconstructItinerary{
  fun findItinerary(tickets: Array<Array<String>>): List<String> {
    tickets.sortWith(compareByDescending{it[1]})
    val nexts = HashMap<String, Stack<String>>()
    for((fr, to) in tickets){
      if(nexts[fr] == null)
        nexts[fr] = Stack()
      nexts[fr]!!.add(to)
    }

    val ret = mutableListOf<String>()
    fun dfs(node: String){
      while(nexts.getOrDefault(node, Stack()).isNotEmpty())
        dfs(nexts[node]!!.pop())          // smaller first
      ret.add(node)     //reverse order
    }
    dfs("JFK")
    return ret.reversed()
  }

  @Test
  fun test1(){
    assertEquals(listOf("JFK", "MUC", "LHR", "SFO", "SJC"), findItinerary(arrayOf(
        arrayOf("MUC", "LHR"),
        arrayOf("JFK", "MUC"),
        arrayOf("SFO", "SJC"),
        arrayOf("LHR", "SFO")
    )))
  }

  @Test
  fun test2(){
    assertEquals(listOf("JFK","NRT","JFK","KUL"), findItinerary(arrayOf(
        arrayOf("JFK", "KUL"),
        arrayOf("JFK", "NRT"),
        arrayOf("NRT", "JFK")
    )))
  }

}